Save and load the current planned hyperjump in the save file

JonBooth78 opened the pull request:
The planned hyperspace route, with all stops is saved along with the save game, and restored on load.

impaktor commented:
This fixes #5178 ?

JonBooth78 commented:
@impaktor , only half of it; I'll get onto the rest later - thanks for pointing it out

Web-eWorks reviewed (commented):
Especially given that you're planning on serializing more sector view state, I'd recommend that you marshal the serialized data through a `Serializer:Register()` call in `map-sector-view.lua` instead of registering the serialization callback in `hyperjump-planner.lua`.

The hyperjump planner / sector view split is slightly unconventional / unfinished in that the hyperjump planner is not quite a re-usable class that the sector map owns an instance of (yet), but it's close enough that I'd recommend applying that same mental model.

This is not strictly required, but it will make it easier to clean up / finish encapsulating the hyperjump planner widget in the future.

Web-eWorks commented:
Hmm... the more I look at this code the more I think the serialization should be entirely handled in `map-sector-view.lua`. 99% of this code is manipulating the C++ sector view object itself, and the only thing that involves the hyperjump planner widget is a one-off call in an onGameStart event to `hyperjumpPlanner.buildRouteList()`.

JonBooth78 commented:
@Web-eWorks , I think your right; I also suspect that the call to `hyperjumpPlanner.buildRouteList()` is also redundant as I think it gets called on entry to the screen; I just wanted to be super-safe.

I felt that as this was all to do with the route-planner, it made sense to live in here, although I agree, it really has little to do with that element, which is a UX one.

I'll have a crack at moving it to the suggested place and add the other missing bits of #5187

JonBooth78 commented:
@Web-eWorks , ready for re-review, all seems to be complete and working, with a bonus bug-fix thrown in at the end :-)

JonBooth78 commented:
oh snap, take that back found a bug

JonBooth78 commented:
OK, fixed that bug (doesn't look like I introduced it, oddly) and then clang formatted a file I touched on the way through. @Web-eWorks now ready to go

Web-eWorks commented on data/pigui/views/map-sector-view.lua:
```diff
@@ -17,6 +17,8 @@ local Color = _G.Color
 local ui = require 'pigui'
 local layout = require 'pigui.libs.window-layout'
 
+local Serializer = require 'Serializer'
```
Teensy tiny nitpick but we generally prefer to require "core" modules like Serializer, Game, Event etc. at the top of the file before other requires. In this case, I'd move the `require 'Serializer'` statement to just after (or before, if you want to sort semi-alphabetically) `require 'SystemPath'`.

Web-eWorks commented on src/SectorView.cpp:
```diff
@@ -24,9 +20,13 @@
 #include "lua/LuaRef.h"
 #include "lua/LuaTable.h"
 #include "matrix4x4.h"
+#include <assert.h>
+#include <cmath>
+#include <sstream>
```
Since clang-format is already nitpicking this, I'd move these includes down by one line so they're contiguous with the `#include <unordered_set>` statement and separated from `#include "matrix4x4.h".

Web-eWorks reviewed (approved):
Two very minor tiny nitpicks but the code looks good otherwise. Haven't had time to test this branch, so it's provisionally approved pending a playtest sometime in the next 24 hours.

Web-eWorks commented:
> OK, fixed that bug (doesn't look like I introduced it, oddly) and then clang formatted a file I touched on the way through.

This is quite the ancient bug oddly enough (4 years old!) but was completely benign up until yesterday when I merged #5724. I had the feeling something was odd with the `onEnterSystem` callback signature but rationalized it away due to lack of time. Thanks for catching that!
